libraries/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -88,108 +142,123 @@ static bool _packetMatch( const IotLink_t * pSubscriptionLink,
 
 /*-----------------------------------------------------------*/
 
-static bool _topicMatch( const IotLink_t * pSubscriptionLink,
-                         void * pMatch )
+static bool _matchTerminalWildcards( const char * pTopicFilter,
+                                     uint16_t nameIndex,
+                                     uint16_t filterIndex,
+                                     uint16_t topicNameLength,
+                                     uint16_t topicFilterLength )
 {
-    bool status = false;
-    uint16_t nameIndex = 0, filterIndex = 0;
+    bool status = false, terminalChar = false;
 
-    /* Because this function is called from a container function, the given link
-     * must never be NULL. */
-    IotMqtt_Assert( pSubscriptionLink != NULL );
+    /* Determine if the terminal character is reached for both topic name and topic
+     * filter for the '#' wildcard. */
+    terminalChar = ( nameIndex == topicNameLength - 1 ) && ( filterIndex == topicFilterLength - 3 );
 
-    _mqttSubscription_t * pSubscription = IotLink_Container( _mqttSubscription_t,
-                                                             pSubscriptionLink,
-                                                             link );
-    _topicMatchParams_t * pParam = ( _topicMatchParams_t * ) pMatch;
+    if( terminalChar == true )
+    {
+        /* Determine if the topic filter ends with the '#' wildcard. */
+        status = ( pTopicFilter[ filterIndex + 1 ] == '/' ) && ( pTopicFilter[ filterIndex + 2 ] == '#' );
```

Comment:
Are you guaranteed that filterIndex + 2 is inbounds for pTopicFilter?

Reply:
Yes.  The check at line 155 `( filterIndex == topicFilterLength - 3 )` requires a length of at least 3 to pass.